[ASM] Add event rules version in telemetry

NachoEchevarria opened the pull request:
## Summary of changes

In the WAF related telemetry metrics, we are not currently sending the rules version, only the WAF version. All the other libraries already send that and we have been requested to send this metric as well.

For now, the tag is only sent in non-RASP ASM metrics. RASP metrics will include this tag in the near future but the it is still not defined.

## Reason for change

## Implementation details

## Test coverage

## Other details
<!-- Fixes #{issue} -->

<!--  ⚠️ Note: where possible, please obtain 2 approvals prior to merging. Unless CODEOWNERS specifies otherwise, for external teams it is typically best to have one review from a team member, and one review from apm-dotnet. Trivial changes do not require 2 reviews. -->

andrewlock commented:
## Benchmarks Report for appsec :snail:

Benchmarks for #6208 compared to [master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd): 
* 1 benchmarks are **faster**, with geometric mean 1.194
* 1 benchmarks are **slower**, with geometric mean 1.118
* 2 benchmarks have **fewer** allocations


The following thresholds were used for comparing the benchmark speeds:
* Mann–Whitney U test with statistical test for significance of **5%**
* Only results indicating a difference greater than **10%** and **0.3 ns** are considered.

Allocation changes below **0.5%** are ignored.


### Benchmark details


<details><summary><strong>Benchmarks.Trace.Asm.AppSecBodyBenchmark</strong> - Slower :warning: Same allocations :heavy_check_mark:



#### Slower :warning: in #6208

|  Benchmark         | diff/base | Base Median (ns) | Diff Median (ns) | Modality   |
|:----------|-----------:|-----------:|--------:|--------:|
| Benchmarks.Trace.Asm.AppSecBodyBenchmark.ObjectExtractorMoreComplexBody&#8209;net6.0 | 1.118 | 2,953.88 | 3,301.50 | 



#### Faster :tada: in #6208

|  Benchmark         | base/diff | Base Median (ns) | Diff Median (ns) | Modality   |
|:----------|-----------:|-----------:|--------:|--------:|
| Benchmarks.Trace.Asm.AppSecBodyBenchmark.ObjectExtractorMoreComplexBody&#8209;net472 | 1.194 | 4,494.64 | 3,765.07 | 

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`AllCycleSimpleBody`|net6.0|75.2μs|129ns|498ns|0.0746| 0|0|6.01&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`AllCycleSimpleBody`|netcoreapp3.1|63μs|80.3ns|300ns|0.0938| 0|0|6.95&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`AllCycleSimpleBody`|net472|49.9μs|58ns|225ns|1.31| 0|0|8.33&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`AllCycleMoreComplexBody`|net6.0|79μs|50.2ns|188ns|0.12| 0|0|9.51&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`AllCycleMoreComplexBody`|netcoreapp3.1|69.5μs|64.7ns|251ns|0.139| 0|0|10.36&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`AllCycleMoreComplexBody`|net472|56.9μs|86.4ns|335ns|1.88| 0.0285|0|11.85&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`ObjectExtractorSimpleBody`|net6.0|145ns|0.179ns|0.692ns|0.00397| 0|0|280&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`ObjectExtractorSimpleBody`|netcoreapp3.1|202ns|0.294ns|1.14ns|0.00367| 0|0|272&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`ObjectExtractorSimpleBody`|net472|167ns|0.0968ns|0.375ns|0.0446| 0|0|281&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`ObjectExtractorMoreComplexBody`|net6.0|2.95μs|1.79ns|6.68ns|0.0534| 0|0|3.78&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`ObjectExtractorMoreComplexBody`|netcoreapp3.1|4.04μs|1.53ns|5.72ns|0.0505| 0|0|3.69&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`ObjectExtractorMoreComplexBody`|net472|4.5μs|5.33ns|20.6ns|0.602| 0.00448|0|3.8&nbsp;KB|
|#6208|`AllCycleSimpleBody`|net6.0|74.5μs|65.4ns|245ns|0.0743| 0|0|6.01&nbsp;KB|
|#6208|`AllCycleSimpleBody`|netcoreapp3.1|62μs|53ns|198ns|0.0926| 0|0|6.95&nbsp;KB|
|#6208|`AllCycleSimpleBody`|net472|49.7μs|57.4ns|222ns|1.32| 0|0|8.34&nbsp;KB|
|#6208|`AllCycleMoreComplexBody`|net6.0|79.8μs|101ns|390ns|0.119| 0|0|9.51&nbsp;KB|
|#6208|`AllCycleMoreComplexBody`|netcoreapp3.1|70μs|72.4ns|271ns|0.14| 0|0|10.37&nbsp;KB|
|#6208|`AllCycleMoreComplexBody`|net472|55.6μs|122ns|474ns|1.88| 0.0277|0|11.85&nbsp;KB|
|#6208|`ObjectExtractorSimpleBody`|net6.0|145ns|0.191ns|0.741ns|0.00395| 0|0|280&nbsp;B|
|#6208|`ObjectExtractorSimpleBody`|netcoreapp3.1|206ns|0.393ns|1.52ns|0.00377| 0|0|272&nbsp;B|
|#6208|`ObjectExtractorSimpleBody`|net472|167ns|0.181ns|0.675ns|0.0446| 0|0|281&nbsp;B|
|#6208|`ObjectExtractorMoreComplexBody`|net6.0|3.3μs|2.29ns|8.26ns|0.0532| 0|0|3.78&nbsp;KB|
|#6208|`ObjectExtractorMoreComplexBody`|netcoreapp3.1|3.91μs|2.02ns|7.54ns|0.0487| 0|0|3.69&nbsp;KB|
|#6208|`ObjectExtractorMoreComplexBody`|net472|3.76μs|3.28ns|11.8ns|0.602| 0.00563|0|3.8&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.Asm.AppSecEncoderBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`EncodeArgs`|net6.0|38.4μs|18.6ns|71.9ns|0.46| 0|0|32.4&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`EncodeArgs`|netcoreapp3.1|54.2μs|25ns|93.5ns|0.43| 0|0|32.4&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`EncodeArgs`|net472|65.7μs|45.1ns|175ns|5.13| 0.0658|0|32.5&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`EncodeLegacyArgs`|net6.0|72μs|96ns|333ns|0.0365| 0|0|2.14&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`EncodeLegacyArgs`|netcoreapp3.1|106μs|99.4ns|385ns|0| 0|0|2.14&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`EncodeLegacyArgs`|net472|153μs|93.6ns|363ns|0.306| 0|0|2.15&nbsp;KB|
|#6208|`EncodeArgs`|net6.0|37.8μs|23.6ns|91.5ns|0.452| 0|0|32.4&nbsp;KB|
|#6208|`EncodeArgs`|netcoreapp3.1|54.3μs|20.2ns|75.5ns|0.432| 0|0|32.4&nbsp;KB|
|#6208|`EncodeArgs`|net472|65.9μs|34.3ns|133ns|5.16| 0.0658|0|32.5&nbsp;KB|
|#6208|`EncodeLegacyArgs`|net6.0|72.8μs|25.2ns|97.5ns|0| 0|0|2.14&nbsp;KB|
|#6208|`EncodeLegacyArgs`|netcoreapp3.1|108μs|386ns|1.5μs|0| 0|0|2.15&nbsp;KB|
|#6208|`EncodeLegacyArgs`|net472|155μs|82.1ns|318ns|0.309| 0|0|2.15&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.Asm.AppSecWafBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`RunWafRealisticBenchmark`|net6.0|184μs|317ns|1.23μs|0| 0|0|2.44&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`RunWafRealisticBenchmark`|netcoreapp3.1|195μs|241ns|901ns|0| 0|0|2.39&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`RunWafRealisticBenchmark`|net472|209μs|119ns|461ns|0.311| 0|0|2.46&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`RunWafRealisticBenchmarkWithAttack`|net6.0|122μs|72.8ns|282ns|0| 0|0|1.47&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`RunWafRealisticBenchmarkWithAttack`|netcoreapp3.1|129μs|89.6ns|335ns|0| 0|0|1.46&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`RunWafRealisticBenchmarkWithAttack`|net472|138μs|48.2ns|167ns|0.207| 0|0|1.49&nbsp;KB|
|#6208|`RunWafRealisticBenchmark`|net6.0|184μs|128ns|479ns|0| 0|0|2.44&nbsp;KB|
|#6208|`RunWafRealisticBenchmark`|netcoreapp3.1|197μs|121ns|451ns|0| 0|0|2.39&nbsp;KB|
|#6208|`RunWafRealisticBenchmark`|net472|209μs|104ns|401ns|0.313| 0|0|2.46&nbsp;KB|
|#6208|`RunWafRealisticBenchmarkWithAttack`|net6.0|123μs|48.9ns|183ns|0| 0|0|1.47&nbsp;KB|
|#6208|`RunWafRealisticBenchmarkWithAttack`|netcoreapp3.1|130μs|135ns|521ns|0| 0|0|1.46&nbsp;KB|
|#6208|`RunWafRealisticBenchmarkWithAttack`|net472|139μs|40.5ns|151ns|0.209| 0|0|1.49&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.Iast.StringAspectsBenchmark</strong> - Same speed :heavy_check_mark: Fewer allocations :tada:



#### Fewer allocations :tada: in #6208

|  Benchmark         | Base Allocated | Diff Allocated | Change   | Change % |
|:----------|-----------:|-----------:|--------:|--------:|
| Benchmarks.Trace.Iast.StringAspectsBenchmark.StringConcatBenchmark&#8209;net472 | 59.36 KB | 59.01 KB | -352 B | -0.59%
| Benchmarks.Trace.Iast.StringAspectsBenchmark.StringConcatAspectBenchmark&#8209;netcoreapp3.1 | 263.3 KB | 253 KB | -10.3 KB | -3.91%

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`StringConcatBenchmark`|net6.0|62.4μs|667ns|6.64μs|0| 0|0|43.44&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`StringConcatBenchmark`|netcoreapp3.1|53.9μs|287ns|1.49μs|0| 0|0|42.64&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`StringConcatBenchmark`|net472|38.4μs|64.1ns|222ns|0| 0|0|59.36&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`StringConcatAspectBenchmark`|net6.0|309μs|1.7μs|10.7μs|0| 0|0|254.19&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`StringConcatAspectBenchmark`|netcoreapp3.1|347μs|1.87μs|11.1μs|0| 0|0|263.3&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`StringConcatAspectBenchmark`|net472|288μs|6.44μs|62.5μs|0| 0|0|278.53&nbsp;KB|
|#6208|`StringConcatBenchmark`|net6.0|59.4μs|728ns|7.2μs|0| 0|0|43.44&nbsp;KB|
|#6208|`StringConcatBenchmark`|netcoreapp3.1|60.8μs|750ns|7.38μs|0| 0|0|42.64&nbsp;KB|
|#6208|`StringConcatBenchmark`|net472|37.6μs|88.8ns|320ns|0| 0|0|59.01&nbsp;KB|
|#6208|`StringConcatAspectBenchmark`|net6.0|318μs|1.77μs|11.8μs|0| 0|0|254.76&nbsp;KB|
|#6208|`StringConcatAspectBenchmark`|netcoreapp3.1|338μs|1.92μs|15.7μs|0| 0|0|253&nbsp;KB|
|#6208|`StringConcatAspectBenchmark`|net472|287μs|6.24μs|60.8μs|0| 0|0|278.53&nbsp;KB|
</details>

andrewlock commented:
## Benchmarks Report for tracer :snail:

Benchmarks for #6208 compared to [master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd): 
* 2 benchmarks are **faster**, with geometric mean 1.171
* 1 benchmarks are **slower**, with geometric mean 1.164
* All benchmarks have **the same** allocations


The following thresholds were used for comparing the benchmark speeds:
* Mann–Whitney U test with statistical test for significance of **5%**
* Only results indicating a difference greater than **10%** and **0.3 ns** are considered.

Allocation changes below **0.5%** are ignored.


### Benchmark details


<details><summary><strong>Benchmarks.Trace.ActivityBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`StartStopWithChild`|net6.0|7.66μs|38.2ns|175ns|0.0119| 0.00396|0|5.42&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`StartStopWithChild`|netcoreapp3.1|9.84μs|53.7ns|294ns|0.0147| 0.00489|0|5.62&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`StartStopWithChild`|net472|16.3μs|49.8ns|186ns|1.03| 0.318|0.0896|6.06&nbsp;KB|
|#6208|`StartStopWithChild`|net6.0|7.82μs|44.1ns|347ns|0.0153| 0.00763|0|5.42&nbsp;KB|
|#6208|`StartStopWithChild`|netcoreapp3.1|9.85μs|51ns|315ns|0.0194| 0.00972|0|5.62&nbsp;KB|
|#6208|`StartStopWithChild`|net472|16.2μs|56.4ns|219ns|1.03| 0.322|0.0965|6.06&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.AgentWriterBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`WriteAndFlushEnrichedTraces`|net6.0|471μs|231ns|865ns|0| 0|0|2.7&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`WriteAndFlushEnrichedTraces`|netcoreapp3.1|621μs|295ns|1.14μs|0| 0|0|2.7&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`WriteAndFlushEnrichedTraces`|net472|893μs|330ns|1.19μs|0.437| 0|0|3.3&nbsp;KB|
|#6208|`WriteAndFlushEnrichedTraces`|net6.0|485μs|173ns|646ns|0| 0|0|2.7&nbsp;KB|
|#6208|`WriteAndFlushEnrichedTraces`|netcoreapp3.1|649μs|317ns|1.23μs|0| 0|0|2.7&nbsp;KB|
|#6208|`WriteAndFlushEnrichedTraces`|net472|838μs|647ns|2.51μs|0.417| 0|0|3.3&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.AspNetCoreBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`SendRequest`|net6.0|196μs|1.1μs|8.59μs|0.212| 0|0|18.45&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`SendRequest`|netcoreapp3.1|225μs|1.32μs|12.4μs|0.218| 0|0|20.61&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`SendRequest`|net472|9.94E&#8209;06ns|9.94E&#8209;06ns|3.85E&#8209;05ns|0| 0|0|0&nbsp;b|
|#6208|`SendRequest`|net6.0|202μs|1.18μs|11.4μs|0.183| 0|0|18.45&nbsp;KB|
|#6208|`SendRequest`|netcoreapp3.1|226μs|1.31μs|11.9μs|0.215| 0|0|20.61&nbsp;KB|
|#6208|`SendRequest`|net472|0.00114ns|0.00049ns|0.0019ns|0| 0|0|0&nbsp;b|
</details>


<details><summary><strong>Benchmarks.Trace.CIVisibilityProtocolWriterBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`WriteAndFlushEnrichedTraces`|net6.0|572μs|2.04μs|7.37μs|0.584| 0|0|41.52&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`WriteAndFlushEnrichedTraces`|netcoreapp3.1|678μs|3.36μs|13.9μs|0.327| 0|0|41.83&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`WriteAndFlushEnrichedTraces`|net472|885μs|3.04μs|11.4μs|8.74| 2.62|0.437|53.37&nbsp;KB|
|#6208|`WriteAndFlushEnrichedTraces`|net6.0|573μs|2.94μs|14.7μs|0.566| 0|0|41.53&nbsp;KB|
|#6208|`WriteAndFlushEnrichedTraces`|netcoreapp3.1|670μs|3.44μs|17.2μs|0.345| 0|0|41.84&nbsp;KB|
|#6208|`WriteAndFlushEnrichedTraces`|net472|872μs|4.36μs|18.5μs|8.45| 2.53|0.422|53.29&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.DbCommandBenchmark</strong> - Faster :tada: Same allocations :heavy_check_mark:



#### Faster :tada: in #6208

|  Benchmark         | base/diff | Base Median (ns) | Diff Median (ns) | Modality   |
|:----------|-----------:|-----------:|--------:|--------:|
| Benchmarks.Trace.DbCommandBenchmark.ExecuteNonQuery&#8209;net6.0 | 1.130 | 1,413.89 | 1,251.08 | 

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`ExecuteNonQuery`|net6.0|1.41μs|1.37ns|5.29ns|0.0142| 0|0|1.02&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`ExecuteNonQuery`|netcoreapp3.1|1.76μs|1.52ns|5.9ns|0.0141| 0|0|1.02&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`ExecuteNonQuery`|net472|2.06μs|2.19ns|8.48ns|0.156| 0|0|987&nbsp;B|
|#6208|`ExecuteNonQuery`|net6.0|1.25μs|1.39ns|5.38ns|0.0144| 0|0|1.02&nbsp;KB|
|#6208|`ExecuteNonQuery`|netcoreapp3.1|1.76μs|1.9ns|7.34ns|0.0132| 0|0|1.02&nbsp;KB|
|#6208|`ExecuteNonQuery`|net472|2.03μs|1.81ns|6.99ns|0.156| 0|0|987&nbsp;B|
</details>


<details><summary><strong>Benchmarks.Trace.ElasticsearchBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`CallElasticsearch`|net6.0|1.25μs|0.447ns|1.67ns|0.0138| 0|0|976&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`CallElasticsearch`|netcoreapp3.1|1.5μs|0.832ns|2.88ns|0.0132| 0|0|976&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`CallElasticsearch`|net472|2.51μs|1.42ns|5.3ns|0.158| 0|0|995&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`CallElasticsearchAsync`|net6.0|1.36μs|0.515ns|1.93ns|0.0131| 0|0|952&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`CallElasticsearchAsync`|netcoreapp3.1|1.58μs|0.55ns|2.06ns|0.0135| 0|0|1.02&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`CallElasticsearchAsync`|net472|2.54μs|1.6ns|5.99ns|0.166| 0|0|1.05&nbsp;KB|
|#6208|`CallElasticsearch`|net6.0|1.18μs|0.651ns|2.52ns|0.0136| 0|0|976&nbsp;B|
|#6208|`CallElasticsearch`|netcoreapp3.1|1.47μs|1.57ns|6.08ns|0.0133| 0|0|976&nbsp;B|
|#6208|`CallElasticsearch`|net472|2.55μs|2.53ns|8.76ns|0.157| 0|0|995&nbsp;B|
|#6208|`CallElasticsearchAsync`|net6.0|1.33μs|0.561ns|2.1ns|0.0133| 0|0|952&nbsp;B|
|#6208|`CallElasticsearchAsync`|netcoreapp3.1|1.59μs|0.995ns|3.85ns|0.0143| 0|0|1.02&nbsp;KB|
|#6208|`CallElasticsearchAsync`|net472|2.59μs|2.73ns|10.6ns|0.167| 0|0|1.05&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.GraphQLBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`ExecuteAsync`|net6.0|1.2μs|0.484ns|1.68ns|0.0132| 0|0|952&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`ExecuteAsync`|netcoreapp3.1|1.61μs|0.434ns|1.62ns|0.0129| 0|0|952&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`ExecuteAsync`|net472|1.8μs|0.733ns|2.64ns|0.145| 0|0|915&nbsp;B|
|#6208|`ExecuteAsync`|net6.0|1.22μs|0.477ns|1.78ns|0.0134| 0|0|952&nbsp;B|
|#6208|`ExecuteAsync`|netcoreapp3.1|1.65μs|0.69ns|2.58ns|0.0124| 0|0|952&nbsp;B|
|#6208|`ExecuteAsync`|net472|1.75μs|0.881ns|3.3ns|0.145| 0|0|915&nbsp;B|
</details>


<details><summary><strong>Benchmarks.Trace.HttpClientBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`SendAsync`|net6.0|4.15μs|2.37ns|8.88ns|0.031| 0|0|2.22&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`SendAsync`|netcoreapp3.1|5.09μs|2.33ns|8.71ns|0.0381| 0|0|2.76&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`SendAsync`|net472|7.76μs|1.9ns|7.35ns|0.498| 0|0|3.15&nbsp;KB|
|#6208|`SendAsync`|net6.0|4.11μs|2.3ns|8.61ns|0.0308| 0|0|2.22&nbsp;KB|
|#6208|`SendAsync`|netcoreapp3.1|5.21μs|1.63ns|6.09ns|0.0364| 0|0|2.76&nbsp;KB|
|#6208|`SendAsync`|net472|7.81μs|2.32ns|8.69ns|0.5| 0|0|3.15&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.ILoggerBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`EnrichedLog`|net6.0|1.55μs|0.725ns|2.62ns|0.0231| 0|0|1.64&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`EnrichedLog`|netcoreapp3.1|2.08μs|0.824ns|3.08ns|0.022| 0|0|1.64&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`EnrichedLog`|net472|2.49μs|1.36ns|5.07ns|0.249| 0|0|1.57&nbsp;KB|
|#6208|`EnrichedLog`|net6.0|1.57μs|1.11ns|4.15ns|0.0226| 0|0|1.64&nbsp;KB|
|#6208|`EnrichedLog`|netcoreapp3.1|2.27μs|1.04ns|3.9ns|0.0216| 0|0|1.64&nbsp;KB|
|#6208|`EnrichedLog`|net472|2.59μs|1.45ns|5.63ns|0.249| 0|0|1.57&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.Log4netBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`EnrichedLog`|net6.0|116μs|153ns|573ns|0.0583| 0|0|4.28&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`EnrichedLog`|netcoreapp3.1|122μs|86.3ns|311ns|0.0606| 0|0|4.28&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`EnrichedLog`|net472|153μs|148ns|572ns|0.69| 0.23|0|4.46&nbsp;KB|
|#6208|`EnrichedLog`|net6.0|120μs|164ns|636ns|0.0595| 0|0|4.28&nbsp;KB|
|#6208|`EnrichedLog`|netcoreapp3.1|123μs|131ns|506ns|0| 0|0|4.28&nbsp;KB|
|#6208|`EnrichedLog`|net472|151μs|146ns|566ns|0.679| 0.226|0|4.46&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.NLogBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`EnrichedLog`|net6.0|3.06μs|0.767ns|2.87ns|0.0307| 0|0|2.2&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`EnrichedLog`|netcoreapp3.1|4.22μs|1.49ns|5.77ns|0.0294| 0|0|2.2&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`EnrichedLog`|net472|4.89μs|2.67ns|10.4ns|0.32| 0|0|2.02&nbsp;KB|
|#6208|`EnrichedLog`|net6.0|3.12μs|1.56ns|6.02ns|0.0297| 0|0|2.2&nbsp;KB|
|#6208|`EnrichedLog`|netcoreapp3.1|4.1μs|1.41ns|5.45ns|0.0287| 0|0|2.2&nbsp;KB|
|#6208|`EnrichedLog`|net472|4.86μs|1.3ns|4.86ns|0.319| 0|0|2.02&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.RedisBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`SendReceive`|net6.0|1.39μs|0.816ns|3.16ns|0.016| 0|0|1.14&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`SendReceive`|netcoreapp3.1|1.79μs|0.844ns|2.92ns|0.0156| 0|0|1.14&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`SendReceive`|net472|2.18μs|1.04ns|4.05ns|0.183| 0.00109|0|1.16&nbsp;KB|
|#6208|`SendReceive`|net6.0|1.37μs|1.12ns|4.33ns|0.0158| 0|0|1.14&nbsp;KB|
|#6208|`SendReceive`|netcoreapp3.1|1.75μs|0.449ns|1.62ns|0.0158| 0|0|1.14&nbsp;KB|
|#6208|`SendReceive`|net472|2.07μs|0.823ns|3.19ns|0.183| 0|0|1.16&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.SerilogBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`EnrichedLog`|net6.0|2.76μs|0.913ns|3.42ns|0.022| 0|0|1.6&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`EnrichedLog`|netcoreapp3.1|3.89μs|2.25ns|8.44ns|0.0215| 0|0|1.65&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`EnrichedLog`|net472|4.46μs|4.67ns|18.1ns|0.322| 0|0|2.04&nbsp;KB|
|#6208|`EnrichedLog`|net6.0|2.69μs|1.09ns|4.22ns|0.0229| 0|0|1.6&nbsp;KB|
|#6208|`EnrichedLog`|netcoreapp3.1|3.95μs|4.3ns|16.6ns|0.0215| 0|0|1.65&nbsp;KB|
|#6208|`EnrichedLog`|net472|4.51μs|4.29ns|16.6ns|0.322| 0|0|2.04&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.SpanBenchmark</strong> - Slower :warning: Same allocations :heavy_check_mark:



#### Slower :warning: in #6208

|  Benchmark         | diff/base | Base Median (ns) | Diff Median (ns) | Modality   |
|:----------|-----------:|-----------:|--------:|--------:|
| Benchmarks.Trace.SpanBenchmark.StartFinishScope&#8209;net6.0 | 1.164 | 492.91 | 573.52 | 



#### Faster :tada: in #6208

|  Benchmark         | base/diff | Base Median (ns) | Diff Median (ns) | Modality   |
|:----------|-----------:|-----------:|--------:|--------:|
| Benchmarks.Trace.SpanBenchmark.StartFinishSpan&#8209;net6.0 | 1.213 | 490.44 | 404.35 | 

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`StartFinishSpan`|net6.0|490ns|0.31ns|1.2ns|0.00809| 0|0|576&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`StartFinishSpan`|netcoreapp3.1|603ns|0.916ns|3.3ns|0.00775| 0|0|576&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`StartFinishSpan`|net472|683ns|0.282ns|1.09ns|0.0916| 0|0|578&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`StartFinishScope`|net6.0|493ns|0.255ns|0.987ns|0.00968| 0|0|696&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`StartFinishScope`|netcoreapp3.1|754ns|4.12ns|22.5ns|0.0093| 0|0|696&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`StartFinishScope`|net472|902ns|0.969ns|3.75ns|0.104| 0|0|658&nbsp;B|
|#6208|`StartFinishSpan`|net6.0|405ns|0.462ns|1.79ns|0.00808| 0|0|576&nbsp;B|
|#6208|`StartFinishSpan`|netcoreapp3.1|617ns|0.408ns|1.58ns|0.00772| 0|0|576&nbsp;B|
|#6208|`StartFinishSpan`|net472|691ns|0.366ns|1.42ns|0.0916| 0|0|578&nbsp;B|
|#6208|`StartFinishScope`|net6.0|574ns|0.357ns|1.38ns|0.00976| 0|0|696&nbsp;B|
|#6208|`StartFinishScope`|netcoreapp3.1|828ns|0.36ns|1.4ns|0.00951| 0|0|696&nbsp;B|
|#6208|`StartFinishScope`|net472|870ns|0.835ns|3.23ns|0.104| 0|0|658&nbsp;B|
</details>


<details><summary><strong>Benchmarks.Trace.TraceAnnotationsBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`RunOnMethodBegin`|net6.0|647ns|0.516ns|2ns|0.00971| 0|0|696&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`RunOnMethodBegin`|netcoreapp3.1|960ns|0.623ns|2.41ns|0.00958| 0|0|696&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/13c270998c92ebdd9432dfd8fedde43b01db6abd)|`RunOnMethodBegin`|net472|1.09μs|0.976ns|3.78ns|0.104| 0|0|658&nbsp;B|
|#6208|`RunOnMethodBegin`|net6.0|654ns|0.363ns|1.41ns|0.00978| 0|0|696&nbsp;B|
|#6208|`RunOnMethodBegin`|netcoreapp3.1|963ns|0.277ns|1.04ns|0.0096| 0|0|696&nbsp;B|
|#6208|`RunOnMethodBegin`|net472|1.12μs|0.313ns|1.13ns|0.104| 0|0|658&nbsp;B|
</details>

andrewlock commented on tracer/src/Datadog.Trace/Telemetry/Collectors/MetricsTelemetryCollectorBase.cs:
```diff
@@ -211,21 +214,29 @@ private async Task AggregateMetricsLoopAsync()
     }
 
-    private string[]? GetTags(string? ns, string[]? metricKeyTags)
+    private string[]? GetTags(string? ns, string[]? metricKeyTags, string name)
```
I'm not a fan of the changes here - it adds a lot of complexity and assumptions where that's just not necessary IMO.

- It assumes that rasp metrics will never record event rules
- It assumes that _only_ rasp metrics will not record event rules

It also highlights for me that a _bunch_ of the ASM metrics are already broken. Anything that _doesn't_ have `waf_version` as the first tag in the metric is broken AFAICT 😬

NachoEchevarria replied:
We include the waf_version in all the ASM metrics, so that would not be a problem. With this new change, all ASM metrics would also include the event rules version but the RASP ones, which include only the WAF version. Anyway, I will test the implementation that you posted in the other comment, which actually decreases complexity. Thanks for the suggestion!

andrewlock replied:
> We include the waf_version in all the ASM metrics

That doesn't appear to be the case, e.g. waf.inputtruncated does not have it 😬 

![image](https://github.com/user-attachments/assets/e5997550-fdc4-441d-b6f4-18f7868ec7c6)


Maybe we should also add a test that asserts that all the ASM metrics _do_ have it (if they are all _supposed_ to have it) 🤔

NachoEchevarria replied:
You are right! I will check these metrics. I would say that they should have the WAF version, but I need to check first. Thanks for pointing this out.

andrewlock commented on tracer/src/Datadog.Trace/Telemetry/Collectors/MetricsTelemetryCollectorBase.cs:
```diff
@@ -211,21 +214,29 @@ private async Task AggregateMetricsLoopAsync()
+
+        if (!isRasp)
+        {
+            metricKeyTags[1] = (_wafAndRulesVersionTags ?? _unknownWafAndRulesVersionTags)[1];
```
This risks accessing outside the array size

andrewlock reviewed (commented):
I think this highlights a few things:

- The existing (pre-this PR) behaviour looks buggy to me - we're overwriting the tags for a _bunch_ of ASM metrics currently AFAICT. The original design assumed that _all_ ASM metrics should have the waf version tagged. That looks fundamentally broken
- I'm not a fan of the changes in `GetTags()` - I think that's piling on the complexity to an already too-complex aspect. 

I think we can _probably_ simplify it to so that the entire implementation is more like this:

```csharp
private string[]? GetTags(string? ns, string[]? metricKeyTags)
{
    if (ns != MetricNamespaceConstants.ASM)
    {
        return metricKeyTags;
    }

    if (metricKeyTags is null)
    {
        return _wafVersionTags ?? _unknownWafVersionTags;
    }

    if (string.Equals(metricKeyTags[0], "waf_version"))
    {
        var wafVersionTag = (_wafVersionTags ?? _unknownWafVersionTags)[0];
        metricKeyTags[0] = wafVersionTag;
    }

    if (metricKeyTags.Length > 1 && string.Equals(metricKeyTags[1], "event_rules_version"))
    {
        var eventRulesTag = (_wafVersionTags ?? _unknownWafVersionTags)[1];
        metricKeyTags[1] = eventRulesTag;
    }

    return metricKeyTags;
}
```

This solves the existing overwriting issue, and adds support for event_rules. It also means we don't need the additional array fields, we just need

```csharp
private static readonly string[] _unknownWafVersionTags = { "waf_version:unknown", "event_rules_version:unknown" };
private string[]? _wafVersionTags; // gets updated with both waf and rules version
```

NachoEchevarria commented:
> I think this highlights a few things:
> 
> * The existing (pre-this PR) behaviour looks buggy to me - we're overwriting the tags for a _bunch_ of ASM metrics currently AFAICT. The original design assumed that _all_ ASM metrics should have the waf version tagged. That looks fundamentally broken
> * I'm not a fan of the changes in `GetTags()` - I think that's piling on the complexity to an already too-complex aspect.
> 
> I think we can _probably_ simplify it to so that the entire implementation is more like this:
> 
> ```cs
> private string[]? GetTags(string? ns, string[]? metricKeyTags)
> {
>     if (ns != MetricNamespaceConstants.ASM)
>     {
>         return metricKeyTags;
>     }
> 
>     if (metricKeyTags is null)
>     {
>         return _wafVersionTags ?? _unknownWafVersionTags;
>     }
> 
>     if (string.Equals(metricKeyTags[0], "waf_version"))
>     {
>         var wafVersionTag = (_wafVersionTags ?? _unknownWafVersionTags)[0];
>         metricKeyTags[0] = wafVersionTag;
>     }
> 
>     if (metricKeyTags.Length > 1 && string.Equals(metricKeyTags[1], "event_rules_version"))
>     {
>         var eventRulesTag = (_wafVersionTags ?? _unknownWafVersionTags)[1];
>         metricKeyTags[1] = eventRulesTag;
>     }
> 
>     return metricKeyTags;
> }
> ```
> 
> This solves the existing overwriting issue, and adds support for event_rules. It also means we don't need the additional array fields, we just need
> 
> ```cs
> private static readonly string[] _unknownWafVersionTags = { "waf_version:unknown", "event_rules_version:unknown" };
> private string[]? _wafVersionTags; // gets updated with both waf and rules version
> ```
> 
> We should also likely add some unit tests that asserts that:
> 
> * _if_ a metric uses `waf_version` then it's _always_ the first element
> * _if_ a metric uses `event_rules_version` then it's _always_ the second element
> * _if_ a metric uses `event_rules_version` then we _always_ have a `waf_version`

Thanks for the feedback! I have updated the code and added the unit tests that you suggested!

andrewlock reviewed (approved):
Great, thanks! 🙂

NachoEchevarria commented:
Thanks for the feedback and suggestions!
